File: src/plnflr/engine/bom.py
"""Bill of materials from a laid-out plan."""

from __future__ import annotations

from decimal import ROUND_HALF_UP, Decimal

from plnflr.domain.models import BillOfMaterials


def waste_pct(area_bought_mm2: int, area_net_mm2: int) -> str:
    if area_net_mm2 <= 0:
        return "0.0"
    ratio = (Decimal(area_bought_mm2 - area_net_mm2) * Decimal(100)) / Decimal(area_net_mm2)
    return str(ratio.quantize(Decimal("0.1"), rounding=ROUND_HALF_UP))


def make_bom(
    *,
    pieces: int,
    full_boards: int,
    boards_per_pack: int | None,
    area_net_mm2: int,
    board_area_mm2: int,
) -> BillOfMaterials:
    bought = full_boards * board_area_mm2
    packs = None
    if boards_per_pack and boards_per_pack > 0:
        packs = (full_boards + boards_per_pack - 1) // boards_per_pack
    return BillOfMaterials(
        pieces=pieces,
        full_boards=full_boards,
        packs=packs,
        area_net_mm2=area_net_mm2,
        area_bought_mm2=bought,
        waste_pct=waste_pct(bought, area_net_mm2),
    )

File: Kernel/Sources/PlnFlrLayout/Bom.swift
import Foundation

func wastePct(areaBoughtMm2: Int, areaNetMm2: Int) -> String {
    guard areaNetMm2 > 0 else { return "0.0" }
    var ratio = Decimal(areaBoughtMm2 - areaNetMm2) * 100 / Decimal(areaNetMm2)
    var rounded = Decimal()
    NSDecimalRound(&rounded, &ratio, 1, .plain)
    return "\(rounded)"
}

func makeBom(
    pieces: Int,
    fullBoards: Int,
    boardsPerPack: Int?,
    areaNetMm2: Int,
    boardAreaMm2: Int,
    label: String = "",
    kind: ZoneKind = .plank
) -> BillOfMaterials {
    let bought = fullBoards * boardAreaMm2
    var packs: Int?
    if let pack = boardsPerPack, pack > 0 {
        packs = (fullBoards + pack - 1) / pack
    }
    return BillOfMaterials(
        areaBoughtMm2: bought,
        areaNetMm2: areaNetMm2,
        fullBoards: fullBoards,
        kind: kind,
        label: label,
        packs: packs,
        pieces: pieces,
        wastePct: wastePct(areaBoughtMm2: bought, areaNetMm2: areaNetMm2)
    )
}
